### PR TITLE
implement comms GC

### LINKED ISSUE #3306: extend GC over comms protocol
## What is the problem being solved

Our #3106 GC plan covers inter-vat intra-SwingSet references, but doesn't address inter-SwingSet references (which travel through the comms vat). The next step is to extend the existing `dropExport`/`retireExport`/`retireImport` plan through the comms vat.

From the kernel's point of view, the comms vat is just another vat, so it sends and expects the same GC deliveries/syscalls as any other. From the comms vat's point of view, the kernel is just another remote (albeit much more synchronous, which reduces the management overhead somewhat).

The task is to build refcount tracking into the comms tables, then react to inbound GC messages (from either the kernel or remotes) by changing these refcounts. This may cause the comms vat to emit outbound GC messages (either to the kernel as a syscall, or to another remote as a protocol message).

## Design

(theses are just some quick protocol notes, I'll edit them into a proper design later)

* incoming drop/retire messages are either "informed" if the sender was fully aware of any cross-on-the-wire re-introductions, or "ignorant" if not
  * comms-kernel messages are always informed
  * comms-comms messages are informed if the inbound message's acknum is equal or greater than the outbound clist record of the last reintroduction, else they are ignorant
* comms will add an `isReachable` flag to all clist entries
  * on import entries, this describes the state of the importer
  * on export entries, it merely records whether we've sent a drop or not, and might not actually be necessary
* comms will add a (reachable, recognizable) refcount tuple to all comms object table entries
  * just like the kernel does:
    * `reachable` is the sum of the `isReachable` flags from all importers, plus one for each resolved promise or auxdata that references the object
    * `recognizable` is the count of all importing clist entries, plus resolved promises and auxdata
* when comms receives an informed `dropExport`, it clears the `isReachable` flag for that importer's clist, which decrements the `reachable` refcount
  * if `reachable` hits zero, comms sends a `dropExport` to the exporter and clears the `isReachable` flag on the exporter's clist entry (again, maybe not really necessary)
* when comms receives an informed `retireExport`, it deletes the importer's clist entry, which decrements the `recognizable` refcount
  * if `recognizable` hits zero, comms sends a `retireExport` to the exporter and deletes the clist entry
* when comms receives an informed(??) `dropImport` (which will always be from an exporter):
  * comms translates the `dropImport` from sender-space to local (comms) space, then deletes the exporting c-list entry
  *  the `reachable` count must already be zero, else the sender of `dropImport` did something wrong
  * comms locates all importers, then deletes the object table entry
  * comms translates a `dropImport` to each importer, then deletes their clist entry
  * comms sends the translated `dropImport` to the importer

If comms receives a `retireExport` or `retireImport` for a ref that is not in the c-list, it should just ignore it. There are two reasons/phases where this might happen. The "normal" one is during a race between the importer sending `retireExport` and the exporter sending `retireImport`. We could choose to track this race in the same way we handle the retirement of promises:
* importer sends `retireExport`, and tracks the (sent seqnum, rref) pair in an ordered list
  * if a message arrives that effectively acks that seqnum, delete the pair: the window for a race has closed
  * if a re-introduction arrives before that point, delete the pair: the race has been superceded by a replacement object
  * if a `retireImport` arrives before that point, ignore it: there was a race, no big deal
  * if a `retireImport` arrives after that point (i.e. neither the clist nor the ordered `retireExport`-sent list knows the rref): this is the weird case, we might decide to kill the connection, or log-but-ignore, or just ignore
 * follow the same pattern when the exporter sends `retireImport`

The algorithm for comms is mostly simpler than the kernel because:
* there are no queued messages: no run-queue, and no per-promise message queues (because everything is immediately pipelined)
* we aren't trying to spread GC actions out among multiple cranks, so we don't need to record the upcoming work in a durable fashion: the equivalents of `maybeFreeKrefs` and the durable `gcActions` set can both be ephemeral
* drops and retires appear in separate messages, so we don't need the `processNextGCAction` code that prioritizes one over the other, and actions won't be negated by an earlier re-introduction

However it's slightly more complex because of the need to distinguish between informed and ignorant inbound messages, and the possibility that we choose to log or kill-connection when a `retireImport`/`retireExport` arrives after we know the race window has closed.

### PR DESCRIPTION
**Pull request by warner**
This patch stack extends the GC protocol to run over the comms layer. I recommend reading them one at a time.

I'm still working on docs improvements and new tests.

@FUDCo I'm posting this early as a draft, in case you have some free time and can start to take a look. I wouldn't be surprised if bugs remain, but the core changes should be stable.

closes #3306

### CODE REVIEW

**Comment by warner** [comment]
I think this is good enough to start reviewing. I'm working on more tests now, but all the existing ones pass without complaint.

**Review by dckc — COMMENTED** [review]
I skimmed each of the commits, FWIW.

**Comment by warner** [comment]
OK, I'm pretty happy with the test coverage now, I found and fixed a few bugs in the process of writing them. I think we're covering all of the various cross-on-the-wire message patterns that could happen.

I've got some new diagrams to add, and prose to accompany it, but I think the code and tests are now ready. I'm going to start up a perf network and use the loadgen scripts to exercise some more integration-ish situations.

File: packages/SwingSet/src/vats/comms/gc-comms.js [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+//
+// Likewise, the GC operations are described from the vat's point of view.
+// When the vat no longer needs an imported object, it calls
+// `syscall.dropImport` to tell the kernel "I'm dropping this thing I
+// imported from you". `syscall.retireImport` is how the vat tells the kernel
+// it can no longer recognize the object and the c-list entry should be
+// deleted. Both messages flow "upstream" from the importing vat to the
+// kernel. Then, if all importers have dropped a kernel object, and nothing
+// else in the kernel can reach it, the kernel will send
+// `dispatch.dropExport` and/or `dispatch.retireExport` to inform the
```
Comment (FUDCo) [inline packages/SwingSet/src/vats/comms/gc-comms.js:30]: It just now occurs to me to ask if `retireExport` subsumes `dropExport`.  That is, does a `retireExport` need to be preceded by a `dropExport` or is that implied automatically?

File: packages/SwingSet/src/vats/comms/state.js [hunk 2]
```diff
@@ -83,10 +87,16 @@ export function makeState(syscall, identifierBase = 0) {
   // p.nextID = $NN  // kernel-facing promise identifier allocation counter (p+NN)
   // c.$kfref = $lref // inbound kernel-facing c-list (o+NN/o-NN/p+NN/p-NN -> loNN/lpNN)
   // c.$lref = $kfref // outbound kernel-facing c-list (loNN/lpNN -> o+NN/o-NN/p+NN/p-NN)
+  // cr.$lref = 1 | <missing> // isReachable flag
```
Comment (FUDCo) [inline packages/SwingSet/src/vats/comms/state.js:90]: In the kernel the reachable flag is encoded in the value of the clist entry, whereas here it appears to be a separate key.  Might be worth a comment talking about the change in design pattern.

**Review by FUDCo — APPROVED** [review]
To the extent that I imagine I understand the kernel GC stuff, this all seems reasonable.  I'm not sure I'm yet able to read closely enough to spot subtle edge-case weirdness, but modulo that I think this looks good.

File: packages/SwingSet/src/vats/comms/gc-comms.js (hunk 1 above)
Comment (warner) [reply packages/SwingSet/src/vats/comms/gc-comms.js:30]: The drop must appear before the retire, but the retire does not imply the drop.

```
state=REACHABLE     <------------------\
 |                                     |
 | dispatch.dropExport                 |
 v                                     |
state=RECOGNIZABLE   ---(new import)->-/
 |                                     |
 | dispatch.retireExport               |
 v                                     |
state=NONE           ---(new import)->-/
```

File: packages/SwingSet/src/vats/comms/state.js (hunk 2 above)
Comment (warner) [reply packages/SwingSet/src/vats/comms/state.js:90]: Yeah, I'm experimenting with a different approach here.. I think I like the separate flag better (easier to manipulate), although it's probably less space-efficient on the DB.

There are a few other differences in approach that I think I'll backport to the kernel implementation eventually. Specifically the way an downstream-travelling `retireImport` simply deletes the upstream-facing c-list entry and marks the ref as `maybeFree`, then lets the main processing loop discover that the exporter's c-list is empty but the recognizable count is still nonzero (which reacts by deleting all the importer's c-list entries and emitting `retireExports`). In the kernel, `disptach.retireImport` deletes the importer's c-list entries immediately. Having written it both ways, I think I like the comms approach better.

**Pull request merged**